File: ios/SsdidWallet/Feature/Credentials/CredentialDetailScreen.swift
import SwiftUI
import SsdidCore

struct CredentialDetailScreen: View {
    @Environment(AppRouter.self) private var router
    @EnvironmentObject private var services: ServiceContainer

    let credentialId: String

    @State private var credential: VerifiableCredential?
    @State private var showRawJson = false
    @State private var showDeleteDialog = false
    @State private var isVerifying = false
    @State private var freshnessRatio: Double = 0.0

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                HStack(spacing: 4) {
                    Button { router.pop() } label: {
                        Image(systemName: "chevron.left")
                            .foregroundStyle(Color.textPrimary)
                            .font(.system(size: 20))
                    }
                    .padding(.leading, 8)

                    Text("Credential Details")
                        .font(.ssdidHeadline)
                        .foregroundStyle(Color.textPrimary)
                }

                Spacer()

                Button { showDeleteDialog = true } label: {
                    Text("Delete")
                        .font(.system(size: 14))
                        .foregroundStyle(Color.danger)
                }
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            if let vc = credential {
                ScrollView {
                    LazyVStack(spacing: 12) {
                        // Main info card
                        VStack(alignment: .leading, spacing: 0) {
                            let isExpired = isCredentialExpired(vc)
                            HStack {
                                Text(vc.type.last ?? "Credential")
                                    .font(.ssdidHeadline)
                                    .foregroundStyle(Color.textPrimary)
                                if vc.type.contains("SdJwtVerifiableCredential") {
                                    Text("SD-JWT VC")
                                        .font(.system(size: 9, weight: .semibold))
                                        .foregroundStyle(Color.ssdidAccent)
                                        .padding(.horizontal, 6)
                                        .padding(.vertical, 2)
                                        .background(Color.accentDim)
                                        .cornerRadius(4)
                                }
                                Spacer()
                                Text(isExpired ? "Expired" : "Valid")
                                    .font(.system(size: 11, weight: .medium))
                                    .foregroundStyle(isExpired ? Color.danger : Color.success)
                                    .padding(.horizontal, 10)
                                    .padding(.vertical, 3)
                                    .background(isExpired ? Color.dangerDim : Color.successDim)
                                    .cornerRadius(4)
                            }

                            Spacer().frame(height: 16)

                            credDetailRow("ID", value: vc.id, mono: true)
                            credDetailRow("Type", value: vc.type.joined(separator: ", "))
                            credDetailRow("Issuer", value: vc.issuer, mono: true)
                            credDetailRow("Issuance Date", value: vc.issuanceDate)
                            if let exp = vc.expirationDate {
                                credDetailRow("Expiration Date", value: exp)
                            }
                        }
                        .ssdidCard()

                        // Credential subject
                        VStack(alignment: .leading, spacing: 0) {
                            Text("CREDENTIAL SUBJECT")
                                .font(.ssdidCaption)
                                .foregroundStyle(Color.textTertiary)

                            Spacer().frame(height: 12)

                            credDetailRow("Subject ID", value: vc.credentialSubject.id, mono: true)

                            if !vc.credentialSubject.claims.isEmpty {
                                Spacer().frame(height: 4)
                                Text("CLAIMS")
                                    .font(.system(size: 10, weight: .medium))
                                    .foregroundStyle(Color.textTertiary)
                                Spacer().frame(height: 8)
                                ForEach(Array(vc.credentialSubject.claims.keys.sorted()), id: \.self) { key in
                                    if let value = vc.credentialSubject.claims[key] {
                                        credDetailRow(key, value: value)
                                    }
                                }
                            } else {
                                Spacer().frame(height: 8)
                                Text("No additional claims")
                                    .font(.system(size: 13))
                                    .foregroundStyle(Color.textTertiary)
                            }
                        }
                        .ssdidCard()

                        // Proof
                        VStack(alignment: .leading, spacing: 0) {
                            Text("PROOF")
                                .font(.ssdidCaption)
                                .foregroundStyle(Color.textTertiary)

                            Spacer().frame(height: 12)

                            credDetailRow("Type", value: vc.proof.type)
                            credDetailRow("Created", value: vc.proof.created)
                            credDetailRow("Purpose", value: vc.proof.proofPurpose)
                            credDetailRow("Verification Method", value: vc.proof.verificationMethod, mono: true)
                        }
                        .ssdidCard()

                        // Bundle freshness badge
                        BundleFreshnessBadge(freshnessRatio: freshnessRatio)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.horizontal, 4)

                        // Verify Credential button
                        Button {
                            guard let vc = credential else { return }
                            let orchestrator = services.verificationOrchestrator
                            isVerifying = true
                            Task {
                                let result = await orchestrator.verify(credential: vc)
                                await MainActor.run {
                                    isVerifying = false
                                    router.push(.verificationResult(result: result))
                                }
                            }
                        } label: {
                            Label(isVerifying ? "Verifying..." : "Verify Credential", systemImage: "checkmark.shield")
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 8)
                        }
                        .disabled(isVerifying)
                        .foregroundStyle(Color.ssdidAccent)
                        .font(.system(size: 14))

                        // Raw JSON toggle
                        Button {
                            showRawJson.toggle()
                        } label: {
                            Text(showRawJson ? "Hide Raw JSON" : "Show Raw JSON")
                                .font(.system(size: 14))
                                .foregroundStyle(Color.ssdidAccent)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 8)
                        }

                        if showRawJson {
                            VStack(alignment: .leading) {
                                Text(prettyPrintJson(vc))
                                    .font(.system(size: 11, design: .monospaced))
                                    .foregroundStyle(Color.textSecondary)
                                    .lineSpacing(4)
                            }
                            .padding(14)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .background(Color.bgSecondary)
                            .cornerRadius(12)
                        }

                        Spacer().frame(height: 20)
                    }
                    .padding(.horizontal, 20)
                }
            } else {
                Spacer()
                ProgressView()
                    .progressViewStyle(CircularProgressViewStyle(tint: Color.ssdidAccent))
                Spacer()
            }
        }
        .background(Color.bgPrimary)
        .alert("Delete Credential", isPresented: $showDeleteDialog) {
            Button("Cancel", role: .cancel) {}
            Button("Delete", role: .destructive) {
                Task {
                    if let vc = credential {
                        try? await services.vault.deleteCredential(credentialId: vc.id)
                        try? await services.credentialRepository.deleteCredential(credentialId: vc.id)
                    }
                    router.pop()
                }
            }
        } message: {
            Text("Are you sure you want to delete this credential? This action cannot be undone.")
        }
        .task {
            // Load the credential from vault
            let all = await services.vault.listCredentials()
            credential = all.first { $0.id == credentialId }

            // Compute bundle freshness for this credential's issuer
            if let vc = credential {
                if let bundle = await services.bundleStore.getBundle(issuerDid: vc.issuer) {
                    freshnessRatio = services.ttlProvider.freshnessRatio(fetchedAt: bundle.fetchedAt)
                }
            }
        }
    }

    @ViewBuilder
    private func credDetailRow(_ label: String, value: String, mono: Bool = false) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label.uppercased())
                .font(.system(size: 10, weight: .medium))
                .foregroundStyle(Color.textTertiary)
            Text(value)
                .font(.system(size: 13, design: mono ? .monospaced : .default))
                .foregroundStyle(Color.textPrimary)
            Spacer().frame(height: 6)
            Divider().background(Color.ssdidBorder)
        }
        .padding(.vertical, 6)
    }

    private func isCredentialExpired(_ vc: VerifiableCredential) -> Bool {
        guard let exp = vc.expirationDate else { return false }
        let formatter = ISO8601DateFormatter()
        guard let date = formatter.date(from: exp) else { return false }
        return Date() > date
    }

    private func prettyPrintJson(_ vc: VerifiableCredential) -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(vc),
              let str = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return str
    }
}

File: android/app/src/main/java/my/ssdid/wallet/feature/credentials/CredentialDetailScreen.kt
package my.ssdid.wallet.feature.credentials

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.time.Instant
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import my.ssdid.wallet.domain.model.VerifiableCredential
import my.ssdid.wallet.domain.revocation.RevocationManager
import my.ssdid.wallet.domain.revocation.RevocationStatus
import my.ssdid.wallet.domain.vault.Vault
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import my.ssdid.wallet.ui.theme.*
import javax.inject.Inject

@HiltViewModel
class CredentialDetailViewModel @Inject constructor(
    private val vault: Vault,
    private val revocationManager: RevocationManager,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    val credentialId: String = savedStateHandle["credentialId"] ?: ""

    private val _credential = MutableStateFlow<VerifiableCredential?>(null)
    val credential = _credential.asStateFlow()

    private val _revocationStatus = MutableStateFlow<RevocationStatus?>(null)
    val revocationStatus = _revocationStatus.asStateFlow()

    private val _deleted = MutableStateFlow(false)
    val deleted = _deleted.asStateFlow()

    init {
        viewModelScope.launch {
            val vc = vault.listCredentials().find { it.id == credentialId }
            _credential.value = vc
            if (vc != null) {
                _revocationStatus.value = revocationManager.checkRevocation(vc)
            }
        }
    }

    fun deleteCredential() {
        viewModelScope.launch {
            vault.deleteCredential(credentialId)
                .onSuccess { _deleted.value = true }
        }
    }
}

@Composable
fun CredentialDetailScreen(
    credentialId: String,
    onBack: () -> Unit,
    viewModel: CredentialDetailViewModel = hiltViewModel()
) {
    val credential by viewModel.credential.collectAsState()
    val deleted by viewModel.deleted.collectAsState()
    val revocationStatus by viewModel.revocationStatus.collectAsState()
    var showRawJson by remember { mutableStateOf(false) }
    var showDeleteDialog by remember { mutableStateOf(false) }

    LaunchedEffect(deleted) {
        if (deleted) onBack()
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding().navigationBarsPadding()
    ) {
        // Header
        Row(
            Modifier
                .fillMaxWidth()
                .padding(start = 8.dp, end = 20.dp, top = 12.dp, bottom = 12.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back", tint = TextPrimary) }
                Spacer(Modifier.width(4.dp))
                Text("Credential Details", style = MaterialTheme.typography.titleLarge)
            }
            TextButton(onClick = { showDeleteDialog = true }) {
                Text("Delete", color = Danger, fontSize = 14.sp)
            }
        }

        credential?.let { vc ->
            LazyColumn(
                Modifier.padding(horizontal = 20.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                // Main info card
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(16.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Column(Modifier.padding(18.dp)) {
                            Row(
                                Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Row(verticalAlignment = Alignment.CenterVertically) {
                                    Text(
                                        vc.type.lastOrNull() ?: "Credential",
                                        style = MaterialTheme.typography.headlineSmall
                                    )
                                    if (vc.type.contains("SdJwtVerifiableCredential")) {
                                        Spacer(Modifier.width(8.dp))
                                        Box(
                                            Modifier
                                                .clip(RoundedCornerShape(4.dp))
                                                .background(AccentDim)
                                                .padding(horizontal = 6.dp, vertical = 2.dp)
                                        ) {
                                            Text(
                                                "SD-JWT VC",
                                                fontSize = 9.sp,
                                                color = Accent,
                                                fontWeight = FontWeight.SemiBold
                                            )
                                        }
                                    }
                                }
                                val isExpired = vc.expirationDate?.let {
                                    try { Instant.now().isAfter(Instant.parse(it)) } catch (_: Exception) { false }
                                } ?: false
                                val isRevoked = revocationStatus == RevocationStatus.REVOKED
                                val statusUnknown = revocationStatus == RevocationStatus.UNKNOWN

                                val statusText = when {
                                    isRevoked -> "Revoked"
                                    isExpired -> "Expired"
                                    statusUnknown -> "Status Unknown"
                                    else -> "Valid"
                                }
                                val statusColor = when {
                                    isRevoked || isExpired -> Danger
                                    statusUnknown -> Warning
                                    else -> Success
                                }
                                val statusBg = when {
                                    isRevoked || isExpired -> DangerDim
                                    statusUnknown -> WarningDim
                                    else -> SuccessDim
                                }

                                Box(
                                    Modifier
                                        .clip(RoundedCornerShape(4.dp))
                                        .background(statusBg)
                                        .padding(horizontal = 10.dp, vertical = 3.dp)
                                ) {
                                    Text(
                                        statusText,
                                        fontSize = 11.sp,
                                        color = statusColor,
                                        fontWeight = FontWeight.Medium
                                    )
                                }
                            }
                            Spacer(Modifier.height(16.dp))

                            CredDetailRow("ID", vc.id, mono = true)
                            CredDetailRow("Type", vc.type.joinToString(", "))
                            CredDetailRow("Issuer", vc.issuer, mono = true)
                            CredDetailRow("Issuance Date", vc.issuanceDate)
                            vc.expirationDate?.let { CredDetailRow("Expiration Date", it) }
                        }
                    }
                }

                // Credential subject claims
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(16.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Column(Modifier.padding(18.dp)) {
                            Text("CREDENTIAL SUBJECT", style = MaterialTheme.typography.labelMedium)
                            Spacer(Modifier.height(12.dp))

                            CredDetailRow("Subject ID", vc.credentialSubject.id, mono = true)

                            if (vc.credentialSubject.claims.isNotEmpty()) {
                                Spacer(Modifier.height(4.dp))
                                Text("CLAIMS", style = MaterialTheme.typography.labelSmall)
                                Spacer(Modifier.height(8.dp))
                                vc.credentialSubject.claims.forEach { (key, value) ->
                                    CredDetailRow(key, value)
                                }
                            } else {
                                Spacer(Modifier.height(8.dp))
                                Text("No additional claims", fontSize = 13.sp, color = TextTertiary)
                            }
                        }
                    }
                }

                // Proof info
                item {
                    Card(
                        Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(12.dp),
                        colors = CardDefaults.cardColors(containerColor = BgCard)
                    ) {
                        Column(Modifier.padding(18.dp)) {
                            Text("PROOF", style = MaterialTheme.typography.labelMedium)
                            Spacer(Modifier.height(12.dp))
                            CredDetailRow("Type", vc.proof.type)
                            CredDetailRow("Created", vc.proof.created)
                            CredDetailRow("Purpose", vc.proof.proofPurpose)
                            CredDetailRow("Verification Method", vc.proof.verificationMethod, mono = true)
                        }
                    }
                }

                // Raw JSON toggle
                item {
                    TextButton(
                        onClick = { showRawJson = !showRawJson },
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Text(
                            if (showRawJson) "Hide Raw JSON" else "Show Raw JSON",
                            color = Accent,
                            fontSize = 14.sp
                        )
                    }
                }

                if (showRawJson) {
                    item {
                        Card(
                            Modifier.fillMaxWidth(),
                            shape = RoundedCornerShape(12.dp),
                            colors = CardDefaults.cardColors(containerColor = BgSecondary)
                        ) {
                            val prettyJson = Json { prettyPrint = true }
                            val rawJson = prettyJson.encodeToString(vc)
                            Text(
                                rawJson,
                                modifier = Modifier.padding(14.dp),
                                fontFamily = FontFamily.Monospace,
                                fontSize = 11.sp,
                                color = TextSecondary,
                                lineHeight = 16.sp
                            )
                        }
                    }
                }

                // Bottom spacer
                item { Spacer(Modifier.height(20.dp)) }
            }
        } ?: run {
            Box(
                Modifier
                    .fillMaxSize()
                    .padding(32.dp),
                contentAlignment = Alignment.Center
            ) {
                CircularProgressIndicator(color = Accent, strokeWidth = 2.dp)
            }
        }
    }

    // Delete confirmation dialog
    if (showDeleteDialog) {
        AlertDialog(
            onDismissRequest = { showDeleteDialog = false },
            title = { Text("Delete Credential", color = TextPrimary) },
            text = { Text("Are you sure you want to delete this credential? This action cannot be undone.", color = TextSecondary) },
            confirmButton = {
                TextButton(onClick = {
                    showDeleteDialog = false
                    viewModel.deleteCredential()
                }) {
                    Text("Delete", color = Danger)
                }
            },
            dismissButton = {
                TextButton(onClick = { showDeleteDialog = false }) {
                    Text("Cancel", color = TextSecondary)
                }
            },
            containerColor = BgElevated
        )
    }
}

@Composable
private fun CredDetailRow(label: String, value: String, mono: Boolean = false) {
    Column(Modifier.padding(vertical = 6.dp)) {
        Text(label.uppercase(), style = MaterialTheme.typography.labelSmall)
        Spacer(Modifier.height(2.dp))
        Text(
            value,
            fontSize = 13.sp,
            color = TextPrimary,
            fontFamily = if (mono) FontFamily.Monospace else FontFamily.Default
        )
        Spacer(Modifier.height(6.dp))
        HorizontalDivider(color = Border)
    }
}
